File: conformance/faults/lostreply/python/lostreply.py
"""Lose the reply to one committed request frame in Python client fixtures.

LostReply wraps a transport with exchange_frame(frame) -> bytes. The wrapped
exchange runs to completion, so the service has processed the request and
answered; the caller receives a connection-loss error instead of the reply. It
fires once, for the first frame whose JSON "method" member equals the method.

preload_environment builds the environment for a subprocess whose frames pass
through the native IPC library, loading the shared lostreply.c shim.
"""
import json
import os
import threading


def _connection_lost(message):
    try:
        from abstraction.ipc import DISCONNECTED, FrameError
    except ImportError:
        return ConnectionResetError(message)
    return FrameError(DISCONNECTED, message)


def _method(frame):
    try:
        value = json.loads(frame)
    except (TypeError, ValueError, UnicodeDecodeError):
        return None
    return value.get("method") if isinstance(value, dict) else None


class LostReply:
    """Transport wrapper; forwards call_scope and write_frame to the inner transport."""

    def __init__(self, inner, method="Submit", error=_connection_lost):
        self.inner, self.method, self._error = inner, method or "Submit", error
        self._lock = threading.Lock()
        self.fired = False
        self.discarded = 0

    def exchange_frame(self, frame):
        reply = self.inner.exchange_frame(frame)
        with self._lock:
            if not self.fired and _method(frame) == self.method:
                self.fired, self.discarded = True, len(reply)
                raise self._error("fixture discarded the committed " + self.method + " reply")
        return reply

    def write_frame(self, frame):
        return self.inner.write_frame(frame)

    def call_scope(self):
        return self.inner.call_scope()


def preload_environment(shim, marker=None, method="Submit", base=None):
    """Environment for a native-IPC subprocess with the lostreply.c shim preloaded."""
    env = dict(os.environ if base is None else base)
    preload = env.get("LD_PRELOAD")
    env["LD_PRELOAD"] = str(shim) + (":" + preload if preload else "")
    env["OA_LOST_REPLY_METHOD"] = method
    if marker is not None:
        env["OA_LOST_REPLY_MARKER"] = str(marker)
    return env

File: conformance/faults/lostreply/go/lostreply.go
// Package lostreply loses the reply to one committed request frame for Go
// client fixtures. The wrapped exchange runs to completion, so the service has
// processed the request and answered; the caller receives ErrLostReply instead
// of the reply. It fires once, for the first frame whose JSON "method" member
// equals the configured method.
package lostreply

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

// ErrLostReply wraps io.ErrUnexpectedEOF, the error a truncated reply produces.
var ErrLostReply = fmt.Errorf("lostreply: fixture discarded the committed reply: %w", io.ErrUnexpectedEOF)

// Exchanger is the frame transport shape generated Go clients accept.
type Exchanger interface {
	ExchangeFrame([]byte) ([]byte, error)
}

type writer interface {
	WriteFrame([]byte) error
}

// Transport discards one reply and passes every other frame through unchanged.
type Transport struct {
	inner     Exchanger
	method    string
	mu        sync.Mutex
	fired     bool
	discarded int
}

// New wraps inner; an empty method selects "Submit".
func New(inner Exchanger, method string) *Transport {
	if method == "" {
		method = "Submit"
	}
	return &Transport{inner: inner, method: method}
}

func (t *Transport) ExchangeFrame(frame []byte) ([]byte, error) {
	reply, err := t.inner.ExchangeFrame(frame)
	if err != nil {
		return reply, err
	}
	t.mu.Lock()
	defer t.mu.Unlock()
	if !t.fired && methodOf(frame) == t.method {
		t.fired, t.discarded = true, len(reply)
		return nil, ErrLostReply
	}
	return reply, nil
}

// WriteFrame forwards one-way frames when the wrapped transport supports them.
func (t *Transport) WriteFrame(frame []byte) error {
	w, ok := t.inner.(writer)
	if !ok {
		return errors.New("lostreply: wrapped transport has no WriteFrame")
	}
	return w.WriteFrame(frame)
}

// Fired reports whether the fault ran and how many reply bytes it discarded.
func (t *Transport) Fired() (bool, int) {
	t.mu.Lock()
	defer t.mu.Unlock()
	return t.fired, t.discarded
}

func methodOf(frame []byte) string {
	var envelope struct {
		Method string `json:"method"`
	}
	if json.Unmarshal(frame, &envelope) != nil {
		return ""
	}
	return envelope.Method
}
